### ISSUE TITLE
Modify all nn.Modules to take in the config as part of their initialization 

### ISSUE THREAD

**Issue by gkielian** [ReaLLMASIC/nanoGPT#43]
This will allow us to simplify and unify the method for setting internal settings within each of these nn.Modules (e.g. the softmax alternatives).
